src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+                         CMSG_SPACE(sizeof(int) * NOTIFY_FD_MAX)) control;
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_control = &control,
+                .msg_controllen = sizeof(control),
+        };
+        struct ucred *ucred;
+        Manager *m = ASSERT_PTR(userdata);
+        Job *j;
+        ssize_t n;
+        unsigned progress;
+        char *p;
+        int r;
+
+        n = recvmsg_safe(fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
+        if (n < 0) {
+                if (ERRNO_IS_TRANSIENT(n))
+                        return 0;
+                return (int) n;
+        }
+
+        cmsg_close_all(&msghdr);
+
+        if (msghdr.msg_flags & MSG_TRUNC) {
+                log_warning("Got overly long notification datagram, ignoring.");
+                return 0;
+        }
+
+        ucred = CMSG_FIND_DATA(&msghdr, SOL_SOCKET, SCM_CREDENTIALS, struct ucred);
+        if (!ucred || ucred->pid <= 0) {
+                log_warning("Got notification datagram lacking credential information, ignoring.");
+                return 0;
+        }
+
+        HASHMAP_FOREACH(j, m->jobs)
+                if (ucred->pid == j->pid)
+                        break;
+
+        if (!j) {
+                log_warning("Got notification datagram from unexpected peer, ignoring.");
+                return 0;
+        }
+
+        buf[n] = 0;
+
+        p = find_line_startswith(buf, "X_IMPORT_PROGRESS=");
+        if (!p)
+                return 0;
+
+        truncate_nl(p);
+
+        r = safe_atou(p, &progress);
+        if (r < 0 || progress > 100) {
+                log_warning("Got invalid percent value, ignoring.");
+                return 0;
+        }
+
+        j->progress_percent = progress;
+        (void) sd_bus_emit_properties_changed(j->manager->bus, j->object_path,
+                                              "org.freedesktop.sysupdate1.Job",
+                                              "Progress", NULL);
+
+        log_debug("Got percentage from worker " PID_FMT ": %u%%", j->pid, j->progress_percent);
+        return 0;
+}
+
+static int manager_new(Manager **ret) {
+        _cleanup_(manager_unrefp) Manager *m = NULL;
+        static const union sockaddr_union sa = {
+                .un.sun_family = AF_UNIX,
+                .un.sun_path = "/run/systemd/sysupdate/notify",
+        };
+        int r;
+
+        assert(ret);
+
+        m = new(Manager, 1);
+        if (!m)
+                return -ENOMEM;
+
+        m->targets = hashmap_new(&string_hash_ops);
+        if (!m->targets)
+                return -ENOMEM;
+
+        m->jobs = hashmap_new(&trivial_hash_ops);
+        if (!m->jobs)
+                return -ENOMEM;
+
+        r = sd_event_default(&m->event);
+        if (r < 0)
+                return r;
+
+        (void) sd_event_set_watchdog(m->event, true);
+
+        r = sd_event_add_signal(m->event, NULL, SIGINT, NULL, NULL);
```

Comment:
in new code please use SD_EVENT_SIGNAL_PROCMASK right from the start, and drop the manual sigprocmask() in the main() func

Reply:
Looks like I can't do anything about SIGCHILD, though, for `sd_event_add_child`? So I cannot drop the manual `sigprocmask` AFAICT